overviewer_core/src/primitives/overlay-bounds.c

```diff
@@ -0,0 +1,249 @@
+                    // Exception set automatically
+                    free(bounds);
+                    self->bounds = NULL;
+                    return 1;
+                }
+
+                // Parse colorpy into a c-struct.
+                if (!PyArg_ParseTuple( colorpy, "bbbb",
+                                       &bounds[i].r,
+                                       &bounds[i].g,
+                                       &bounds[i].b,
+                                       &bounds[i].a)) {
+                    free(bounds);
+                    self->bounds = NULL;
+                    return 1;
+                }
+
+                // Convert condspy to a fast sequence
+                condspy = PySequence_Fast(condspy, "Failed to parse conditions");
+                if(condspy == NULL) {
+                    free(bounds);
+                    self->bounds = NULL;
+                    return 1;
+                }
+
+                // get the number of conditions.
+                bounds[i].numconds = PySequence_Fast_GET_SIZE(condspy);
+                // reserve enough memory for the conditions.
+                cond = calloc(bounds[i].numconds, sizeof(struct Condition));
+                bounds[i].conditions = cond;
+
+                if (bounds[i].conditions == NULL) {
+                    PyErr_SetString(PyExc_MemoryError, "failed to allocate memory");
+                    free(bounds);
+                    self->bounds = NULL;
+                    return 1;
+                }
+
+                // iterate over all the conditions and read them.
+                for (n = 0; n < bounds[i].numconds; n++) {
+                    PyObject *ccond = PySequence_Fast_GET_ITEM(condspy, n);
+                    if(!PyArg_ParseTuple( ccond, "iiii",
+                                          &cond[n].minx,
+                                          &cond[n].minz,
+                                          &cond[n].maxx,
+                                          &cond[n].maxz)){
+                        int x = 0;
+                        for(x = 0; x < bounds_size; x++){
+                            free(bounds[x].conditions);
+                        }
+                        free(bounds);
+                        self->bounds = NULL;
+                        return 1;
+                    }
+                }
+            }
+        }
+    }
+
+    /* setup custom color */
+    self->parent.get_color = get_color;
+
+    return 0;
+}
+
+static void overlay_bounds_finish(void *data, RenderState *state) {
+    /* first free all *our* stuff */
+	RenderPrimitiveBounds* self = (RenderPrimitiveBounds *)data;
+    int i = 0;
+
+    if(self->bounds) {
+        // freeing the nested bounds
+        struct Color * m = self->bounds;
+        for(i = 0; i < self->numcolors; i++){
+            if(m[i].conditions) {
+				free(m[i].conditions);
+			}
+        }
+    }
+
+    if (self->bounds) {
```

Comment:
The same condition in two ifs side by side, join them into one if or fail fast if there is no `self->bounds`.